### DISCUSSION TITLE
Extract a value for a dynamic key

### DISCUSSION THREAD

**Discussion by connec** [colinhacks/zod#2262] [Q&A] [answered]
Hi,

I'm trying to use `zod` to parse values based on a "dynamic schema" of sorts. In particular, I have a `key: string` and want to parse the value for the key from an unknown value.

It's easy to test that the unknown value is an object with the key, and that the key has the expected type:

```ts
const object = z.object({ [key]: z.string() }).parse(unkownValue)
```

However, as far as I can tell there's no way to extract the value we just parsed without assertions. E.g. the following compiles no problem, but I'm wondering if there's a solution that doesn't resort to `!`.

```ts
const value: string = object[key]!
```

Essentially I'm trying to write the following function without `!`:

```ts
function pluck<T>(record: unknown, key: string, parser: z.ZodType<T>): T {
  // The intermediate assignment is necessary to satisfy tsc
  const object: { [k: string]: T } = z.object({ [key]: parser }).parse(record)
  return object[key]!
}
```

As a complete aside, although the `<T>(parser: z.ZodType<T>)` approach is not recommended, the alternative (`<P extends z.ZodTypeAny>(parser: P)`) ends up subverting TypeScript's knowledge that arbitrary object keys might be undefined:

```ts
function pluck<P extends z.ZodAnyType>(record: unknown, key: string, parser: P): z.infer<P> {
  // The intermediate assignment is necessary to satisfy tsc
  const object: { [k: string]: z.infer<P> } = z.object({ [key]: parser }).parse(record)

  // This works 🙌
  return object[key]

  // ... But so does this? 🤔
  return object['nonsense']
}
```

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
function pluck<
    record extends Record<string, any>,
    Key extends keyof record,
    RecordSchema extends z.ZodRecord<z.ZodType<keyof record>, z.ZodTypeAny>,
> (
    record: record,
    key: Key,
    recordSchema: RecordSchema,
): record[ Key ] {
    const object = recordSchema.parse( record )
    return object[ key ] as record[ Key ]
}

const data = pluck(
    { a: 1, b: 2 } as const,
    'a',
    z.record( z.enum( [ 'a', 'b' ] ), z.number() ),
)
type Data = typeof data
// type Data = 1

console.log( data ) // 1

const nonsenseData = pluck(
    { a: 1, b: 2 } as const,
    'nonsense', // Argument of type '"nonsense"' is not assignable to parameter of type '"a" | "b"'.
    z.record( z.enum( [ 'a', 'b' ] ), z.number() ),
)
type NonsenseData = typeof nonsenseData
// type NonsenseData = 1 | 2

console.log( nonsenseData ) // undefined
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
